Logical operators (#56)
javascript-tutorial/pt.javascript.info: 1-js/02-first-steps/11-logical-operators/article.md

# Operadores lógicos

Existem três operadores lógicos em JavaScript: `||` (OU), `&&` (E), `!` (NÃO).

> Embora eles sejam chamados de "lógicos", podem ser aplicados a valores de qualquer tipo, não apenas a boolean. Seus resultados também podem ser de qualquer tipo.

Reviewer: ```suggestion
Embora eles são denominados como "lógicos", podem ser aplicados a valores de qualquer tipo, não apenas ao tipo `boolean`. Seus resultados também podem ser de qualquer tipo.
```

Vamos ver os detalhes.

## || (OU)

O operador "OU" é representado com dois símbolos de linha vertical:

```js
result = a || b;
```

> Em programação clássica, o operador OU é mencionado para manipular apenas valores booleanos. Se qualquer um dos seus argumentos for `true`, ele retorna `true`, se não, retorna `false`.

Reviewer: Complementando a sugestão do @odsantos em [", o operador lógico OU é <s>mencionado</s> utilizado para manipular ..."](https://github.com/javascript-tutorial/pt.javascript.info/pull/56/files#r358993776)

```suggestion
Em programação clássica, o operador OU é utilizado para manipular apenas valores do tipo `boolean`. Se qualquer um dos seus argumentos for `true`, ele retorna `true`, se não, retorna `false`.
```

> Em JavaScript, este operador é um pouco mais útil e poderoso. Mas primeiro, vamos ver o que acontece com valores booleanos.

Reviewer: ```suggestion
Em JavaScript, este operador é um pouco mais útil e poderoso. Mas primeiro, vamos ver o que acontece com valores do tipo `boolean`.
```

alert( false || false ); // false
```

> 

Reviewer: O trecho de código está correto @odsantos.

Como podemos ver, o resultado é sempre `true`, exceto para o caso onde os dois operandos são `false`.

> Se um operando não é um boolean, ele é convertido em um boolean para ser avaliado.

Reviewer: ```suggestion
Se um operando não é um `boolean`, ele é convertido em um `boolean` para ser avaliado.
```

```js run
if (1 || 0) { // funciona como if( true || false)
  alert( 'verdadeiro!');
}
```

Na maioria das vezes, OU `||` é usado dentro de uma expressão `if` para testar se *qualquer* uma das condições dadas é `true`.

Por exemplo:

```js run
let hour = 0;

*!*
if (hour < 10 || hour > 18) {
*/!*
  alert( 'O escritório está fechado.' );
}
```

Nós podemos passar mais condições:

```js run
let hour = 12;
let isWeekend = true;

if (hour < 10 || hour > 18 || isWeekend) {
  alert( 'O escritório está fechado.' ); // é final de semana
}
```

## OU encontra o primeiro valor verdadeiro

A lógica descrita acima é algo clássico. Agora, vamos ver as funcionalidades "extras" do JavaScript.

O algorítmo extendido funciona da seguinte forma.

Dando múltiplos valores encadeados em OU's:

```js
result = value1 || value2 || value3;
```

O operador OU `||` faz o seguinte:

- Avalia os operandos da esquerda para a direita.
- Para cada operando, o converte para boolean. Se o resultado é `true`, para e retorna o valor original daquele operando.
- Se todos os operandos foram avaliados (i.e. todos são `false`), retorna o último operando.

Um valor é retornado na sua forma original, sem conversão.

> Em outras palavras, uma cadeia de OU `"||"` retorna o primeiro valor verdadeiro ou o último se nenhum `true` for encontrado.

Reviewer: ```suggestion
Em outras palavras, uma cadeia de OU `"||"`, retorna o primeiro valor verdadeiro ou o último se nenhum `true` for encontrado.
```

Por exemplo:

```js run
alert( 1 || 0 ); // 1 (1 é verdadeiro)
alert( true || 'não importa o quê' ); // (true é verdadeiro)

alert( null || 1 ); // 1 (1 é o primeiro valor verdadeiro)
alert( null || 0 || 1 ); // 1 (o primeiro valor verdadeiro)
alert( undefined || null || 0 ); // 0 (todos falsos, retorna o último valor)
```

Isso nos mostra algumas utilidades interessantes comparadas ao "puro, clássico, apenas-boolean OU".

1. **Obtendo o primeiro valor verdadeiro de uma lista de variáveis ou expressões.**

    Imagine que temos várias variáveis que podem conter algum dado ou ser `null/undefined`. Como podemos encontrar a primeira com algum dado?

    Nós podemos usar OU `||`:

    ```js run

>     let currentUser = null;

Reviewer 2: - "Nós podemos usar <s>OU</s> OR `||`:"

    let name = currentUser || defaultUser || "unnamed";
    */!*

    alert( name ); // seleciona "John" - o primeiro valor verdadeiro
    ```

    Se ambos `currentUser` e `defaultUser` forem falsos, o resultado será `"unnamed"`.
2. **Avaliação de curto-circuito**

>     Operandos podem não ser apenas valores, mas operações arbitrárias. OU interpreta e testa elas da esquerda para a direita. A avaliação é interrompida quando um valor verdadeiro é encontrado e este valor é retornado. Este processo é chamado de "avaliação de curto-circuito" pois vai o mais curto possível da esquerda para a direita.

Reviewer: Complementando a sugestão

```suggestion
    Operandos podem não ser apenas valores, mas operações arbitrárias. `OR` interpreta e testa elas da esquerda para a direita. A avaliação é interrompida quando um valor verdadeiro é encontrado e este valor é retornado. Este processo é denominado como "avaliação em curto-circuito", pois vai o mais curto possível da esquerda para a direita.
```

    alert(x); // 1
    ```

> 

Reviewer: O trecho de código esta correto.

    Uma atribuição é um caso simples. Outros ##SIDE EFFECTS(???)## podem também estarem envolvidos.

>     Como podemos ver, esse caso é como "uma maneira mais curta de se usar `if`". O primeiro operando é convertido para boolean. Se for false, o segundo operando é avaliado.

Reviewer: Complementando a sugestão do @odsantos em [Dec 17, 2019](https://github.com/javascript-tutorial/pt.javascript.info/pull/56/files#r359009516)

```suggestion
    Como podemos ver, este caso é como "uma maneira mais curta de se usar `if`". O primeiro operando é convertido para o tipo `boolean`. Se for `false`, o segundo operando é avaliado.
```
